// js-viva-api/tests/autocomplete.js
var test = require('tape');
var getBaseUrl = require('./get-base-url.js');
var getQueryParams = require('./get-query-params.js');

module.exports = function(VivaApi) {

  var config = {
    baseUrl: "http://api.vivareal.com/api/1.0",
    apiKey: "183d98b9-fc81-4ef1-b841-7432c610b36e",
    portal: "VR_BR"
  };

  // shortcut function to get an instantiated api object
  function setup() {
    return new VivaApi(config);
  }

  test("Given no parameters, should throw", function(t) {
    var api = setup();

    t.throws(function() {
      api.autocomplete();
    });
    t.end();
  });

  test("Given empty string, should throw", function(t) {
    var api = setup();

    t.throws(function() {
      api.autocomplete("");
    });
    t.end();
  });

  test("Should include the query parameters", function(t) {
    var api = setup();

    var formattedUrl = api.autocomplete("foo");
    var params = getQueryParams(formattedUrl);

    t.equal(params.apiKey, config.apiKey);
    t.equal(params.portal, config.portal);
    t.end();
  });

  test("Test base url generation", function(t) {
    var api = setup();

    var formattedUrl = getBaseUrl(api.autocomplete("foo"));
    t.equal(formattedUrl, config.baseUrl + "/locations/autocomplete");
    t.end();
  });

  test("input inclusion", function(t) {
    var api = setup();
    var params;

    params = getQueryParams(api.autocomplete("foo"));
    t.equal(params.input, "foo");

    params = getQueryParams(api.autocomplete("bogotá"));
    t.equal(params.input, "bogot%C3%A1");

    params = getQueryParams(api.autocomplete("são pau"));
    t.equal(params.input, "s%C3%A3o+pau");
    t.end();
  });

  test("surrounding whitespace should be stripped from the string", function(t) {
    var api = setup();
    var params;

    params = getQueryParams(api.autocomplete("  foo  "));
    t.equal(params.input, "foo");

    params = getQueryParams(api.autocomplete("  bogotá  "));
    t.equal(params.input, "bogot%C3%A1");

    params = getQueryParams(api.autocomplete("  são pau      "));
    t.equal(params.input, "s%C3%A3o+pau");
    t.end();
  });

};
