// js-viva-api/tests/get-property.js
var test = require('tape');
var getBaseUrl = require('./get-base-url.js');
var getQueryParams = require('./get-query-params.js');

module.exports = function(VivaApi) {

  var config = {
    baseUrl: "http://api.vivareal.com/api/1.0",
    apiKey: "183d98b9-fc81-4ef1-b841-7432c610b36e",
    portal: "VR_BR"
  };

  // shortcut function to get an instantiated api object
  function setup() {
    return new VivaApi(config);
  }

  test("Given no parameters, should throw", function(t) {
    var api = setup();

    t.throws(function() {
      api.property();
    });
    t.end();
  });

  test("Given empty string, should throw", function(t) {
    var api = setup();

    t.throws(function() {
      api.property("");
    });
    t.end();
  });

  test("Should include the query parameters", function(t) {
    var api = setup();

    var formattedUrl = api.property("foo");
    var params = getQueryParams(formattedUrl);

    t.equal(params.apiKey, config.apiKey);
    t.equal(params.portal, config.portal);
    t.end();
  });

  test("Test base url generation", function(t) {
    var api = setup();

    var formattedUrl = getBaseUrl(api.property("foo"));
    t.equal(formattedUrl, config.baseUrl + "????");
    t.end();
  });

};
