// js-viva-api/tests/startup.js

var test = require('tape');

module.exports = function(VivaApi) {

  var config = {
    baseUrl: "http://api.vivareal.com/api/1.0",
    apiKey: "183d98b9-fc81-4ef1-b841-7432c610b36e",
    portal: "VR_BR"
  };

  // shortcut function to get an instantiated api object
  function setup() {
    return new VivaApi(config);
  }

  test('Can import successfully', function(t) {
    t.equal(typeof VivaApi, 'function');
    t.end();
  });

  test('Can instantiate given minimum good values', function(t) {
    var api = new VivaApi(config);

    t.ok(api);
    t.equal(typeof api, 'object');
    t.end();
  });

  // What should happen if these correct values are not passed in? Throw an error?
  test('Should throw error if no baseUrl', function(t) {
    var init = function() {
      var api = new VivaApi({
        apiKey: "183d98b9-fc81-4ef1-b841-7432c610b36e",
        portal: "VR_BR" 
      });
    };

    t.throws(init);
    t.end();
  });

  test('Should throw error if baseUrl isnt a string', function(t) {
    var init = function() {
      var api = new VivaApi({
        baseUrl: 42,
        apiKey: "183d98b9-fc81-4ef1-b841-7432c610b36e",
        portal: "VR_BR" 
      });
    };

    t.throws(init);
    t.end();
  });

  test('Should throw error if no apiKey', function(t) {
    var init = function() {
      var api = new VivaApi({
        baseUrl: "http://api.vivareal.com/api/1.0",
        portal: "VR_BR" 
      });
    };

    t.throws(init);
    t.end();
  });

  test('Should throw error if apiKey isnt a string', function(t) {
    var init = function() {
      var api = new VivaApi({
        baseUrl: "http://api.vivareal.com/api/1.0",
        apiKey: 42,
        portal: "VR_BR" 
      });
    };

    t.throws(init);
    t.end();
  });

  test('Should throw error if no portal', function(t) {
    var init = function() {
      var api = new VivaApi({
        baseUrl: "http://api.vivareal.com/api/1.0",
        apiKey: "183d98b9-fc81-4ef1-b841-7432c610b36e"
      });
    };
    t.throws(init);
    t.end();
  });

  test('Should throw error if portal isnt a string', function(t) {
    var init = function() {
      var api = new VivaApi({
        baseUrl: "http://api.vivareal.com/api/1.0",
        apiKey: "183d98b9-fc81-4ef1-b841-7432c610b36e",
        portal: 100
      });
    };
    t.throws(init);
    t.end();
  });

  test('Should throw error if portal isnt one of the valid values', function(t) {
    var init = function() {
      var api = new VivaApi({
        baseUrl: "http://api.vivareal.com/api/1.0",
        apiKey: "183d98b9-fc81-4ef1-b841-7432c610b36e",
        portal: "VR_AR"
      });
    };
    t.throws(init);
    t.end();
  });

  test('Should throw error if token isnt a string', function(t) {
    var init = function() {
      var api = new VivaApi({
        baseUrl: "http://api.vivareal.com/api/1.0",
        apiKey: "183d98b9-fc81-4ef1-b841-7432c610b36e",
        portal: 'VR_BR',
        token: {}
      });
    };
    t.throws(init);
    t.end();
  });

};
